Quorum/apis/block_explorers/source_code.py

```diff
@@ -21,60 +32,36 @@ def _parse_source_code(self) -> None:
         Parses the Solidity source code and stores the contract's AST object.
         """
         source_code_str = "\n".join(self.file_content)
+        tmp_path = Path("tmp.sol")
+
+        with open(tmp_path, 'w') as file:
+            file.write(source_code_str)
 
         try:
-            ast = parser.parse(source_code_str)
-            ast_obj = parser.objectify(ast)
-            contract_name = ast_obj._current_contract.name
-            self._parsed_contract = ast_obj.contracts[contract_name]
+            contract_ast = solcx.compile_files(tmp_path, output_values=["ast"], stop_after="parsing")
+            contract_name = list(contract_ast.keys())[0]
+            self._parsed_contract = contract_ast[contract_name]['ast']
+
         except Exception as e:
             pp.pretty_print(f"Error parsing source code for {self.file_name}: {e}\n"
                             f"Some of the checks will not apply to this contract!!!",
                             pp.Colors.FAILURE)
-
-    def get_constructor(self) -> dict | None:
-        """
-        Retrieves the constructor information from the Solidity contract.
-
-        Returns:
-            (dict | None): Constructor information or None if not found.
-        """
-        if self._parsed_contract:
-            return self._parsed_contract.constructor
-        return None
-
-    def get_dependencies(self) -> list | None:
-        """
-        Retrieves the dependencies from the Solidity contract.
-
-        Returns:
-            (list | None): List of dependencies or None if not found.
-        """
-        if self._parsed_contract:
-            return self._parsed_contract.dependencies
-        return None
-
-    def get_enums(self) -> list | None:
-        """
-        Retrieves the enums from the Solidity contract.
-
-        Returns:
-            (list | None): List of enums or None if not found.
-        """
-        if self._parsed_contract:
-            return self._parsed_contract.enums
-        return None
-
-    def get_events(self) -> list | None:
-        """
-        Retrieves the events from the Solidity contract.
-
-        Returns:
-            (list | None): List of events or None if not found.
-        """
-        if self._parsed_contract:
-            return self._parsed_contract.events
-        return None
+        finally:
+            tmp_path.unlink()
+    
+    def __extract_nodes(self, ast: dict, node_type: ASTOption) -> list:
+        nodes = {}
+        visited = set()
+        for node in ast['nodes']:
+            if node['id'] in visited:
+                continue
+            visited.add(node['id'])
+            if node['nodeType'] == node_type:
+                name = node['name']
+                nodes[name] = node
+            elif 'nodes' in node:
+                nodes.update(self.__extract_nodes(node, node_type))
+        return nodes
```

Comment:
Couple of questions regarding this function:
1. Why `nodes` is a dict but `visited` is a set?
2. Why is there a `node_type` option? Wouldn't you ultimately use both options and then you just traversed the AST twice? You could provide a list of both options and just filter later or even cache both lists and return them.
3. The function is supposed to return a list but it returns a dict (not really a question)

Reply:
1. visited is key to tracking the visited nodes and nodes are the results. 
2. we load all lazy to the cache so i am not sure what you expect here.
3. return dict.

Reply:
This might be a small improvement, but he meant cache as like when you look for functions, you retrieve state_variables as well and vice versa.

Reply:
O(2 (V + E)) = O(V + E)

Reply:
So why did you add the chaching in first place?
Do you expect it to be called with the same source code more than once?
What happens if we call it with a different source code?

Reply:
Not sure what you mean, the chaching i added is to prevent execution multiple times on the same call. i.e. calling the get functions multiple times. it does better to make sure it will run only once instead of many times for the same results.

Reply:
Code as code, the more you look at it, the more comments you have 😆